### ISSUE TITLE
400 requires authentication received for /wsapi/list_emails

### ISSUE THREAD

**Issue by jrgm** [mozilla/persona#1449]
This is not a good bug report, since I don't really have STR (yet). But, I just had a series of 400
requires authentication errors on /wsapi/list_emails.  By series, I mean I switched back to a winxp
that I had been using late Sunday. On 3 of the browsers (chrome, opera, ie8), where I had whatever
state from that time, I just went to beer, click signin, am prompted for an email address, enter a
know secondary that I had created today, prompted for password, and got the 400 response:

```
We are very sorry, there has been an error!
Please close this window and try again.
Action: Authenticating User
Network Info: GET: /wsapi/list_emails
Response Code - 400
Response Text: Bad Request: requires authentication
Error Type: Bad Request: requires authentication
```

In chrome devtools, I have no cookies for anything, but in localStorage for diresworb.org, there is this bit of state:

key: emailToUserID, value: {"abcabdab+s1@gmail.com":3938721}
key: emails,        value: {}
key: usersComputer, value: {
     "3938721":{"state":"seen","updated":"Sun Apr 15 2012 20:33:43 GMT-0700 (Pacific Daylight Time)"},
     "3957423":{"state":"seen","updated":"Tue Apr 17 2012 11:55:58 GMT-0700 (Pacific Daylight Time)"}
     }

And I know from other info that 3957423 -> "abcabdab+r41@gmail.com", which was the user I was trying to sign-in.

Marking this 5-star while I figure out if I can reproduce. I've suspended that VM with all three browsers still up and in an error state, so if there's any other info you want, let me know.

**Comment by jrgm** [comment]
Okay, still don't know what's causing it, but I reproduced on another VM (that also had some stale info), and have the full network trace leading to the error, and the before state of localStorage. Digging that out now.

**Comment by jrgm** [comment]
Here's the abbreviated sequence of requests:

00:00:13.138 0.252 395 (2965) GET  304 text/html        http://beta.myfavoritebeer.org/
00:00:13.834 0.095 377 173    GET  200 application/json http://beta.myfavoritebeer.org/api/whoami
00:00:15.735 0.076 360 225    GET  200 text/html        https://diresworb.org/communication_iframe

00:00:16.454 0.286 347 270    GET  200 text/html        https://diresworb.org/sign_in
00:00:17.797 0.081 669 421    GET  200 application/json https://diresworb.org/wsapi/session_context

00:00:30.791 0.219 699 261    GET  200 application/json https://diresworb.org/wsapi/address_info?email=abcabdab%2Br41%40gmail.com
00:00:36.715 0.565 838 275    POST 200 application/json https://diresworb.org/wsapi/authenticate_user
00:00:37.286 0.021 708 296    GET  400 text/plain       https://diresworb.org/wsapi/list_emails

**Comment by lloyd** [comment]
Ok, finally I have a theory of the bug.  To understand it, we need to dig into how `node-client-sessions` works a bit - the library we use for sessioning which is implemented using encrypted, signed cookies.  `node-client-sessions` embeds cookie creation time and cookie validity period into the cookie body, and typically the Expires: field of the cookie is the same as the validity period.

One trick we introduced this train along with the "public computer" support is to extend cookie duration at the point that a user says "this is my computer".  That means changing the embedded validity period, but not other elements of the cookie.  Another behavior client-sessions right now is that when you change the cookie duration, it does not neccesarily update the creation time of the cookie.  So all this interacts in the following way to cause the issue:
1. user hits some browserid API, not authenticated
2. a cookie is set with "now" as the creation time and a default validity period: 7 days
3. the user goes away for an hour
4. the user returns to the computer and signs in with an account that hasn't been used on the computer 
5. upon authentication the server updates the cookie, because this session is "ephemeral" (they haven't used this account yet and confimed they own it), the session validity is set to 1 hour, but the creation time is not updated.

now the front end goes and tries to make calls that require authentication, but because creation time + validity < now, they are not authenticated.

@jrgm - is this a plausible cause of the problems you've seen?

As far as a fix, I think any time we change the cookie in certain ways, we must update the creation time.  I'll add the low lever support to update creation time to `node-client-sessions`, review all changes to cookies, and try to craft a minimal fix.

**Comment by lloyd** [comment]
STR (untested): 
1. enter private browsing mode or open an "icognito" window or clear all stored data for `https://diresworb.org`
2. visit `http://beta.myfavoritebeer.org`, click sign in, then cancel the window
3. let an hour pass
4. click sign in again on myfavoritebeer and try to log in

**Comment by jrgm** [comment]
That sounds about right.

**Comment by jrgm** [comment]
Working for me on stage train-2012.04.11-7. Continuing with some other browsers, etc.

**Comment by jrgm** [comment]
verified fixed for stage train-2012.04.11-8

**Comment by jrgm** [comment]
just noting still fixed in stage train-2012.04.27
